README
Class-User-DBI

A User class: Login credentials and roles.

Through a DBIx::Connector object, this module models a "User" class, with
login credentials, and access roles.  Login credentials include a passphrase,
and optionally per user IP whitelisting.

The module is designed to simplify user logins, and basic administrative user
maintenance.  Passphrases are salted with a 512 bit random salt (unique per
user) using a cryptographically strong random number generator, and converted
to a SHA2-512 digest before being stored in the database.  All subsequent
passphrase validation checks test against the salt and passphrase SHA2 hash.

IP whitelists may be maintained per user.  If a user is set to require an IP
check, then the user validates only if his passphrase authenticates AND his
IP is found in the whitelist associated with his user id.

Users may be given a role, roles may have privileges.  Users may have multiple
"domains" (localities).  See the POD for the full documentation, and
./examples for an example script.
./scripts contains a database configuration script.

DEPENDENCIES

This module requires List::MoreUtils, DBIx::Connector, and 
Authen::Passphrase::SaltedSHA512.

It also requires a database back-end.  The test suite will use DBD::SQLite,
but it has also been tested with DBD::mysql.  None of these dependencies could
be considered light-weight.  The dependency chain of this module is
indicative of the difficulty in assuring cryptographically strong random
salt generation, reliable SHA2-512 hashing of passphrases, fork-safe database
connectivity, and transactional commits for inserts and updates spanning
multiple tables. 

CONFIGURATION

The database needs seven tables.  They can be automatically generated by running
the configuration script provided in the ./scripts directory.

Please read the module's POD for a more thorough explanation and tips on 
configuring the database.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Class::User::DBI

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Class-User-DBI

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Class-User-DBI

    CPAN Ratings
        http://cpanratings.perl.org/d/Class-User-DBI

    Search CPAN
        http://search.cpan.org/dist/Class-User-DBI/


LICENSE AND COPYRIGHT

Copyright (C) 2012 David Oswald

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Perl module Class::User::DBI

0.08 2012-11-02
  - "Changes" now CPAN::Changes::Spec compliant.
  - t/09-changes.t added to verify compliance.

0.07 2012-09-27
  - Eliminated some code that prevented running under pre-5.10 Perls (we
    should be ok through 5.8.x now, again).

0.06 2012-08-30
  - Changed 'id' datatype to BIGINT in user_ips table to resolve issue where
    mysql and sqlite have different ways of designating an unsigned int type.
    Failure to designate an unsigned int (or larger) for IP storage results
    in unreliable round-trip.

0.05 2012-08-21
  - Added setter and getter for ip_required field for users.
  - load_profile() now also includes an ip_required field.
  - Minor documentation updates.

0.04 2012-08-15
  - Class::User::DBI->list_users now returns a sorted list, and includes all
    fields except salt and passphrase hash.

0.03 2012-07-23
  - Corrected some POD typos.
  - Fixed a spelling mistake in the META_MERGE segment within Makefile.PL
  - POD enhancements regarding configuration, examples, and explanation of
    the class heirarchy.

0.02 2012-07-22
  - Major rewrite.
  - Added Domains, UserDomains, Roles, Privileges, and RolePrivileges classes.
  - Full POD rewrite.
  - Many bugfixes.
  - Added ./examples/
  - Reconfigured ./scripts/cudbi-configdb
  - Many new test.
  - Fully operational now.
  - API has stabilized.

0.01_002 2012-06-11
  - Corrected some example code in the POD.
  - Added cudbi-configdb script to scripts/ for automatic database setup.
  - Added a note to the POD about cudbi-configdb.
  - Still beta quality.

0.01_001 2012-06-09
  - First version.
  - Beta.







lib/Class/User/DBI/DB.pm
## no critic (RCS,VERSION,POD)
package Class::User::DBI::DB;

use strict;
use warnings;
use 5.008;

use Exporter;
our @ISA       = qw( Exporter );     ## no critic (ISA)
our @EXPORT    = qw( db_run_ex );    ## no critic (export)
our @EXPORT_OK = qw(
  %USER_QUERY
  %PRIV_QUERY
  %DOM_QUERY
  %ROLE_QUERY
  %RP_QUERY
  %UD_QUERY
  _db_run_ex
);

use Carp;

our $VERSION = '0.08';
# $VERSION = eval $VERSION;    ## no critic (eval)

# ---------------- SQL queries for Class::User::DBI --------------------------

our %USER_QUERY = (
    SQL_get_valid_ips   => 'SELECT ip FROM user_ips WHERE userid = ?',
    SQL_get_credentials => 'SELECT salt, password, ip_required '
      . 'FROM users WHERE userid = ?',
    SQL_exists_user => 'SELECT userid FROM users WHERE userid = ?',
    SQL_load_profile =>
      'SELECT userid, username, email, role, ip_required FROM users WHERE userid = ?',
    SQL_add_ips    => 'INSERT INTO user_ips ( userid, ip ) VALUES( ?, ? )',
    SQL_delete_ips => 'DELETE FROM user_ips WHERE userid = ? AND ip = ?',
    SQL_add_user => 'INSERT INTO users ( userid, salt, password, ip_required, '
      . 'username, email, role ) VALUES( ?, ?, ?, ?, ?, ?, ? )',
    SQL_delete_user     => 'DELETE FROM users WHERE userid = ?',
    SQL_delete_user_ips => 'DELETE FROM user_ips WHERE userid = ?',
    SQL_set_email       => 'UPDATE users SET email = ? WHERE userid = ?',
    SQL_set_username    => 'UPDATE users SET username = ? WHERE userid = ?',
    SQL_set_ip_required => 'UPDATE users SET ip_required = ? where userid = ?',
    SQL_update_password =>
      'UPDATE users SET salt = ?, password = ? WHERE userid = ?',
    SQL_list_users => 'SELECT userid, username, email, role, ip_required FROM users ORDER BY userid',
    SQL_get_role   => 'SELECT role FROM users WHERE userid = ?',
    SQL_get_ip_required => 'SELECT ip_required FROM users WHERE userid = ?',
    SQL_is_role    => 'SELECT role FROM users WHERE userid = ? AND role = ?',
    SQL_set_role   => 'UPDATE users SET role = ? WHERE userid = ?',
    SQL_configure_db_users => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS users (
        userid      VARCHAR(24)           NOT NULL DEFAULT '',
        salt        CHAR(128)             DEFAULT NULL,
        password    CHAR(128)             DEFAULT NULL,
        ip_required TINYINT(1)            NOT NULL DEFAULT '1',
        username    VARCHAR(40)           DEFAULT NULL,
        email       VARCHAR(320)          DEFAULT NULL,
        role        VARCHAR(24)           DEFAULT NULL,
        PRIMARY KEY( userid )
    )
END_SQL
    SQL_configure_db_user_ips => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS user_ips (
        userid      VARCHAR(24)           NOT NULL DEFAULT '',
        ip          BIGINT                NOT NULL DEFAULT '0',
        PRIMARY KEY ( userid, ip )
    )
END_SQL
);

#------------ Queries for Class::User::DBI::Privileges -----------------------

our %PRIV_QUERY = (
    SQL_configure_db_cud_privileges => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS cud_privileges (
        privilege   VARCHAR(24)           NOT NULL,
        description VARCHAR(40)           NOT NULL DEFAULT '',
        PRIMARY KEY (privilege)
    )
END_SQL
    SQL_exists_privilege =>
      'SELECT privilege FROM cud_privileges WHERE privilege = ?',
    SQL_add_privileges =>
      'INSERT INTO cud_privileges ( privilege, description ) VALUES ( ?, ? )',
    SQL_delete_privileges => 'DELETE FROM cud_privileges WHERE privilege = ?',
    SQL_get_privilege_description =>
      'SELECT description FROM cud_privileges WHERE privilege = ?',
    SQL_set_privilege_description =>
      'UPDATE cud_privileges SET description = ? WHERE privilege = ?',
    SQL_list_privileges => 'SELECT * FROM cud_privileges',
);

#----------------- Queries for Class::User::DBI::Domains ---------------------

our %DOM_QUERY = (
    SQL_configure_db_cud_domains => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS cud_domains (
        domain      VARCHAR(24)           NOT NULL,
        description VARCHAR(40)           NOT NULL DEFAULT '',
        PRIMARY KEY (domain)
    )
END_SQL
    SQL_exists_domain => 'SELECT domain FROM cud_domains WHERE domain = ?',
    SQL_add_domains =>
      'INSERT INTO cud_domains ( domain, description ) VALUES ( ?, ? )',
    SQL_delete_domains => 'DELETE FROM cud_domains WHERE domain = ?',
    SQL_get_domain_description =>
      'SELECT description FROM cud_domains WHERE domain = ?',
    SQL_set_domain_description =>
      'UPDATE cud_domains SET description = ? WHERE domain = ?',
    SQL_list_domains => 'SELECT * FROM cud_domains',
);

#----------------- Queries for Class::User::DBI::Roles ---------------------

our %ROLE_QUERY = (
    SQL_configure_db_cud_roles => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS cud_roles (
        role        VARCHAR(24)           NOT NULL,
        description VARCHAR(40)           NOT NULL DEFAULT '',
        PRIMARY KEY (role)
    )
END_SQL
    SQL_exists_role => 'SELECT role FROM cud_roles WHERE role = ?',
    SQL_add_roles =>
      'INSERT INTO cud_roles ( role, description ) VALUES ( ?, ? )',
    SQL_delete_roles => 'DELETE FROM cud_roles WHERE role = ?',
    SQL_get_role_description =>
      'SELECT description FROM cud_roles WHERE role = ?',
    SQL_set_role_description =>
      'UPDATE cud_roles SET description = ? WHERE role = ?',
    SQL_list_roles => 'SELECT * FROM cud_roles',
);

# ----------------- Queries for Class::User::DBI::RolePrivileges ------------

our %RP_QUERY = (
    SQL_configure_db_cud_roleprivs => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS cud_roleprivs (
        role        VARCHAR(24)           NOT NULL DEFAULT '',
        privilege   VARCHAR(24)           NOT NULL DEFAULT '',
        PRIMARY KEY (role,privilege)
    )
END_SQL
    SQL_exists_priv =>
      'SELECT privilege FROM cud_roleprivs WHERE role = ? AND privilege = ?',
    SQL_add_priv =>
      'INSERT INTO cud_roleprivs ( role, privilege ) VALUES ( ?, ? )',
    SQL_delete_privileges =>
      'DELETE FROM cud_roleprivs WHERE role = ? AND privilege = ?',
    SQL_list_privileges => 'SELECT privilege FROM cud_roleprivs WHERE role = ?',
);

# ----------------- Queries for Class::User::DBI::UserDomains ------------

our %UD_QUERY = (
    SQL_configure_db_cud_userdomains => << 'END_SQL',
    CREATE TABLE IF NOT EXISTS cud_userdomains (
        userid      VARCHAR(24)           NOT NULL DEFAULT '',
        domain      VARCHAR(24)           NOT NULL DEFAULT '',
        PRIMARY KEY (userid,domain)
    )
END_SQL
    SQL_exists_domain =>
      'SELECT domain FROM cud_userdomains WHERE userid = ? AND domain = ?',
    SQL_add_domain =>
      'INSERT INTO cud_userdomains ( userid, domain ) VALUES ( ?, ? )',
    SQL_delete_domains =>
      'DELETE FROM cud_userdomains WHERE userid = ? AND domain = ?',
    SQL_list_domains => 'SELECT domain FROM cud_userdomains WHERE userid = ?',
);

# ------------------------------ Functions -----------------------------------

# Prepares and executes a database command using DBIx::Connector's 'run'
# method.  Pass bind values as 2nd+ parameter(s).  If the first bind-value
# element is an array ref, bind value params will be executed in a loop,
# dereferencing each element's list upon execution:
# $self->_db_run_ex( 'SQL GOES HERE', @execute_params ); .... OR....
# $self->_db_run_ex(
#     'SQL GOES HERE',
#     [ first param list ], [ second param list ], ...
# );

sub db_run_ex {
    my ( $conn, $sql, @ex_params ) = @_;
    croak ref($conn) . ' is not a DBIx::Connector.'
      if !$conn->isa('DBIx::Connector');
    my $sth = $conn->run(
        fixup => sub {
            my $sub_sth = $_->prepare($sql);

            # Pass an array of arrayrefs if execute() is to be called in a loop.
            if ( @ex_params && ref( $ex_params[0] ) eq 'ARRAY' ) {
                foreach my $param (@ex_params) {
                    $sub_sth->execute( @{$param} );
                }
            }
            else {
                $sub_sth->execute(@ex_params);
            }
            return $sub_sth;
        }
    );
    return $sth;
}

1;

__END__

=head1 NAME

Class::User::DBI::DB - An internal-use class for the various 
C<Class::User::DBI::*> classes.

=head1 VERSION

Version 0.08

=head1 SYNOPSIS

    my $sql = Class::User::DBI::DB::$USER_QUERY{SQL_some_query};
    
    db_run_ex( $connector, $sql, @parameters );
    # or
    db_run_ex( $connector, $sql, ( [ @params1 ], [ @params2 ] ) );


=head1 DESCRIPTION


This package contains all of the SQL queries for the classes in this 
distribution.

There is also one subroutine, intended for use by the classes.  It handles
database queries through the DBIx::Connector object.

This package is generally not intended for external consumption.  However, if
your database flavor doesn't like this distribution's SQL, you may be relieved
to find out that this is the B<only> place you need to look for SQL problems.
All of the distribution's SQL is here; even table creation.

=head1 EXPORT

Exports (by request) hashes containing the SQL queries for each class.  The
hashes are given names that represent abbreviated versions of the classes they
support.

By  default it also exports C<db_run_ex>, which is a utility function 
facilitating database queries through the connector object.

=head1 SUBROUTINES/METHODS


=head2  db_run_ex

    db_run_ex( $connector, $sql, @parameters );

    # or

    db_run_ex( $connector, $sql, ( [ @params1 ], [ @params2 ] ) );

Pass a connector object, an SQL query, and a list of bind values.  For multiple
C<execute()> calls, pass an array of array refs instead of a flat array.  Each
anonymous array holds the bind values for a single C<< $sth->execute() >> call.

=head1 DEPENDENCIES

The dependencies for this module are the same as for L<Class::User::DBI>, from
this same distribution.  Refer to the documentation in that module for a full
description.


=head1 CONFIGURATION AND ENVIRONMENT

All of the SQL for this entire distribution is found in the 
L<Class::User::DBI::DB> module.  Any adjustments required to suit your database
engine may be made here.  This module's SQL is known to run unaltered with 
SQLite and MySQL.

=head1 DIAGNOSTICS

If you find that your particular database engine is not playing nicely with the
test suite from this module, it may be necessary to provide the database login 
credentials for a test database using the same engine that your application 
will actually be using.  You may do this by setting C<$ENV{CUDBI_TEST_DSN}>,
C<$ENV{CUDBI_TEST_DATABASE}>, C<$ENV{CUDBI_TEST_USER}>, 
and C<$ENV{CUDBI_TEST_PASS}>.

Currently the test suite tests against a SQLite database since it's such a
lightweight dependency for the testing.  The author also uses this module
with several MySQL databases.  As you're configuring your database, providing
its credentials to the tests and running the test scripts will offer really 
good diagnostics if some aspect of your database tables proves to be at odds 
with what this module needs.


=head1 INCOMPATIBILITIES

This module has only been tested on MySQL and SQLite database engines.  If you
are successful in using it with other engines, please send me an email detailing
any additional configuration changes you had to make so that I can document
the compatibility, and improve the documentation for the configuration process.

=head1 BUGS AND LIMITATIONS

=head1 AUTHOR


David Oswald, C<< <davido at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-class-user-dbi at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Class-User-DBI>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Class::User::DBI::DB


You can also look for information at:

=over 4

=item * Class-User-DBI on Github

L<https://github.com/daoswald/Class-User-DBI.git>

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Class-User-DBI>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Class-User-DBI>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Class-User-DBI>

=item * Search CPAN

L<http://search.cpan.org/dist/Class-User-DBI/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 David Oswald.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


lib/Class/User/DBI/Roles.pm
## no critic (RCS,VERSION)
package Class::User::DBI::Roles;

use strict;
use warnings;

use Carp;

use Class::User::DBI::DB qw( db_run_ex  %ROLE_QUERY );

our $VERSION = '0.08';
# $VERSION = eval $VERSION;    ## no critic (eval)

# Two tables: One table is role, description.  Second table is role, privilege.
# Second table allows duplicate roles, but not duplicate role/priv.
# This may be two classes: One for the role/description table, and one for the
# roles/privileges table.

# Class methods.

sub new {
    my ( $class, $conn ) = @_;
    my $self = bless {}, $class;
    croak 'Constructor called without a DBIx::Connector object.'
      if !ref $conn || !$conn->isa('DBIx::Connector');
    $self->{_db_conn} = $conn;
    $self->{roles}    = {};
    return $self;
}

sub configure_db {
    my ( $class, $conn ) = @_;
    croak 'Must provide a valid constructor object.'
      if !ref $conn || !$conn->isa('DBIx::Connector');
    $conn->run(
        fixup => sub {
            $_->do( $ROLE_QUERY{SQL_configure_db_cud_roles} );
        }
    );
    return 1;
}

# Object methods.

sub _db_conn {
    my $self = shift;
    return $self->{_db_conn};
}

# Usage:
# $role->exists_role( $some_role );
# returns 0 or 1.
sub exists_role {
    my ( $self, $role ) = @_;
    croak 'Must pass a defined value in role test.'
      if !defined $role;
    croak 'Must pass a non-empty value in role test.'
      if !length $role;
    return 1 if exists $self->{roles}{$role};
    my $sth = db_run_ex( $self->_db_conn, $ROLE_QUERY{SQL_exists_role}, $role );
    my $result = defined $sth->fetchrow_array;
    $self->{roles}{$role}++ if $result;    # Cache the result.
    return $result;
}

# Usage:
# $role->add_roles( [ qw( role description ) ], [...] );
# Returns the number of roles actually added.

sub add_roles {
    my ( $self, @roles ) = @_;
    my @roles_to_insert =
      grep { ref $_ eq 'ARRAY' && $_->[0] && !$self->exists_role( $_->[0] ) }
      @roles;

    # Set undefined descriptions to q{}.
    foreach my $role_bundle (@roles_to_insert) {

        # This change is intended to propagate back to @roles_to_insert.
        $role_bundle->[1] = q{} if !$role_bundle->[1];
    }
    my $sth =
      db_run_ex( $self->_db_conn, $ROLE_QUERY{SQL_add_roles},
        @roles_to_insert );
    return scalar @roles_to_insert;
}

# Deletes all roles in @roles (if they exist).
# Silent if non-existent. Returns the number of roles actually deleted.
sub delete_roles {
    my ( $self, @roles ) = @_;
    my @roles_to_delete;
    foreach my $role (@roles) {
        next if !$role || !$self->exists_role($role);
        push @roles_to_delete, [$role];
        delete $self->{roles}{$role};    # Remove it from the cache too.
    }
    my $sth = db_run_ex( $self->_db_conn, $ROLE_QUERY{SQL_delete_roles},
        @roles_to_delete );
    return scalar @roles_to_delete;
}

# Gets the description for a single role.  Must specify a valid role.
sub get_role_description {
    my ( $self, $role ) = @_;
    croak 'Must specify a role.'
      if !defined $role;
    croak 'Specified role must exist.'
      if !$self->exists_role($role);
    my $sth =
      db_run_ex( $self->_db_conn, $ROLE_QUERY{SQL_get_role_description},
        $role );
    return ( $sth->fetchrow_array )[0];
}

# Pass a role and a new description.  All parameters required.  Description
# of q{} deletes the description.
sub set_role_description {
    my ( $self, $role, $description ) = @_;
    croak 'Must specify a role.'
      if !defined $role;
    croak 'Specified role doesn\'t exist.'
      if !$self->exists_role($role);
    croak 'Must specify a description (q{} is ok too).'
      if !defined $description;
    my $sth = db_run_ex( $self->_db_conn, $ROLE_QUERY{SQL_set_role_description},
        $description, $role );
    return 1;
}

# Returns an array of pairs (AoA).  Pairs are [ role, description ],...
sub fetch_roles {
    my $self  = shift;
    my $sth   = db_run_ex( $self->_db_conn, $ROLE_QUERY{SQL_list_roles} );
    my @roles = @{ $sth->fetchall_arrayref };
    return @roles;
}

1;

__END__

=head1 NAME

Class::User::DBI::Roles - A Roles class.

=head1 VERSION

Version 0.08

=head1 SYNOPSIS

Through a DBIx::Connector object, this module models a "Roles" class, used for
Roles Based Access Control.  L<Class::User::DBI> allows each user to have a 
single role.  L<Class::User::DBI::RolePrivileges> allows each role to have 
multiple privileges.  And so goes the heirarchy: A user has a role, and a role 
has privileges.

    # Set up a connection using DBIx::Connector:
    # MySQL database settings:

    my $conn = DBIx::Connector->new(
        'dbi:mysql:database=cudbi_tests, 'testing_user', 'testers_pass',
        {
            RaiseError => 1,
            AutoCommit => 1,
        }
    );


    # Now we can play with Class::User::DBI::Roles

    # Set up a 'roles' table in the database.
    Class::User::DBI::Roles->configure_db( $conn );
    
    my $r = Class::User::DBI::Roles->new( $conn );

    $r->add_roles( 
        [ 'office', 'Office users'  ],
        [ 'admin',  'Administrator' ],
    );

    print "Role exists." if $r->exists_role( 'office' );

    my @roles = $r->fetch_roles;
    foreach my $role ( @roles ) {
        my( $name, $description ) = @{$role};
        print "$name => $description\n";
    }

    print "Description for 'office' role: ", 
          $r->get_role_description( 'office' );
    
    $r->set_role_description( 'office', 'Office staff' );
    
    $r->delete_roles( 'office' ); # Pass a list for multiple deletes.


=head1 DESCRIPTION

This is a maintenance class facilitating the creation, deletion, and testing of
roles that are compatible with Class::User::DBI's roles, and 
Class::User::DBI::RolePrivileges roles.

A common usage is to configure a database table, and then add a few roles
along with their descriptions.  Think of roles as similar to Unix 'groups'.
Then use Class::User::DBI::Priviliges to create a few privileges.  Next use
Class::User::DBI::RolePrivileges to associate one or more privileges with a
given role.  Finally, use Class::User::DBI to associate a role with one or
more users.

L<Class::User::DBI::RolePrivileges> provides accessors to query whether a role
has a given privilege.  L<Class::User::DBI> provides a means to query whether
a user has a role.  As a convenience, Class::User::DBI also provides an accessor
to the RolePrivileges class for quickly drilling through from the user to the
privilege.

=head1 EXPORT

Nothing is exported.  There are many object methods, and three class methods,
described in the next section.


=head1 SUBROUTINES/METHODS


=head2  new
(The constructor -- Class method.)

    my $role_obj = Class::User::DBI::Roles->new( $connector );

Creates a role object that can be manipulated to set and get roles from the
database's 'cud_roles' table.  Pass a DBIx::Connector object as a parameter.  
Throws an exception if it doesn't get a valid DBIx::Connector.


=head2  configure_db
(Class method)

    Class::User::DBI::Roles->configure_db( $connector );

This is a class method.  Pass a valid DBIx::Connector as a parameter.  Builds
a minimal database table in support of the Class::User::DBI::Roles 
class.

The table created will be C<cud_roles>.

=head2 add_roles

    $r->add_roles( [ 'office', 'The office staff' ], ... );

Add one or more roles.  Each role must be bundled along with its description in
an array ref.  Pass an AoA for multiple roles, or just an aref for a single
role/description pair.

It will drop requests to add roles that already exist.

Returns a count of roles added, which may be less than the number passed if one
already existed.

=head2 delete_roles

    $r->delete_roles( 'office', 'sales' );

Deletes from the database all roles specified.  Return value is the number of
roles actually deleted, which may be less than the number of roles requested if
any of the requested roles didn't exist in the database to begin with.


=head2 exists_role

    print "Role exists." if $r->exists_role( 'office' );

Returns true if a given role exists, and false if not.

=head2 fetch_roles

    foreach my $role ( $r->fetch_roles ) {
        print "$role->[0] = $role->[1]\n";
    }
    
Returns an array of array refs.  Each array ref contains the role's name and its
description as the first and second elements, respectively.

An empty list means there are no roles defined.

=head2 get_role_description

    my $description = $r->get_role_description( 'office' );
    
Returns the description for a given role.  Throws an exception if the role 
doesn't exist, so be sure to test with C<< $r->exists_role( 'office' ) >> first.

=head2 set_role_description

    $r->set_role_description( 'office', 'New description for office role.' );

Sets a new description for a given role.  If the role doesn't exist in the
database, if not enough parameters are passed, or if any of the params are
C<undef>, an exception will be thrown.  To update a role by giving it a blank
description, pass an empty string as the description.


=head1 DEPENDENCIES

The dependencies for this module are the same as for L<Class::User::DBI>, from
this same distribution.  Refer to the documentation in that module for a full
description.


=head1 CONFIGURATION AND ENVIRONMENT

Please refer to the C<configure_db()> class method for this module for a
simple means of creating the table that supports this class.

All SQL for this distribution is contained in the L<Class::User::DBI::DB> 
module.


=head1 DIAGNOSTICS

If you find that your particular database engine is not playing nicely with the
test suite from this module, it may be necessary to provide the database login 
credentials for a test database using the same engine that your application 
will actually be using.  You may do this by setting C<$ENV{CUDBI_TEST_DSN}>,
C<$ENV{CUDBI_TEST_DATABASE}>, C<$ENV{CUDBI_TEST_USER}>, 
and C<$ENV{CUDBI_TEST_PASS}>.

Currently the test suite tests against a SQLite database since it's such a
lightweight dependency for the testing.  The author also uses this module
with several MySQL databases.  As you're configuring your database, providing
its credentials to the tests and running the test scripts will offer really 
good diagnostics if some aspect of your database tables proves to be at odds 
with what this module needs.


=head1 INCOMPATIBILITIES

This module has only been tested on MySQL and SQLite database engines.  If you
are successful in using it with other engines, please send me an email detailing
any additional configuration changes you had to make so that I can document
the compatibility, and improve the documentation for the configuration process.

=head1 BUGS AND LIMITATIONS

=head1 AUTHOR


David Oswald, C<< <davido at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-class-user-dbi at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Class-User-DBI>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Class::User::DBI::Roles


You can also look for information at:

=over 4

=item * Class-User-DBI on Github

L<https://github.com/daoswald/Class-User-DBI.git>

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Class-User-DBI>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Class-User-DBI>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Class-User-DBI>

=item * Search CPAN

L<http://search.cpan.org/dist/Class-User-DBI/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 David Oswald.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Makefile.PL
## no critic (RCS,VERSION,interpolation)
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Class::User::DBI',
    AUTHOR             => q{David Oswald <davido [@at] cpan [d.o.t] org>},
    VERSION_FROM       => 'lib/Class/User/DBI.pm',
    ABSTRACT_FROM      => 'lib/Class/User/DBI.pm',
    LICENSE            => 'perl',
    MIN_PERL_VERSION   => '5.008000',
    PL_FILES           => {},
    EXE_FILES          => ['scripts/cudbi-configdb'],
    CONFIGURE_REQUIRES => { 'ExtUtils::MakeMaker' => '6.62', },
    BUILD_REQUIRES     => {
        'Test::More'      => '0.98',
        'Test::Exception' => 0,
        'List::MoreUtils' => 0,
        'DBIx::Connector' => 0,
        'DBD::SQLite'     => 0,
    },
    PREREQ_PM => {
        'Authen::Passphrase::SaltedSHA512' => '0.05',
        'DBIx::Connector'                  => 0,
        'List::MoreUtils'                  => 0,
    },
    META_MERGE => {
        provides => {
            'Class::User::DBI' => {
                file    => 'lib/Class/User/DBI.pm',
                version => '0.08',
            },
        },
        version   => '0.08',
        resources => {
            repository => {
                web  => 'http://github.com/daoswald/Class-User-DBI.git',
                url  => 'git://github.com/daoswald/Class-User-DBI.git',
                type => 'git',
            },
        },
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'Class-User-DBI-*' },
);
